pip/pip-369.md

```diff
@@ -0,0 +1,89 @@
+# PIP-371: Support to specify message listeners in TableView constructor
+
+# Background knowledge
+
+### TableView 
+Ref: https://pulsar.apache.org/docs/next/concepts-clients/#tableview
+
+### TableView Listeners
+TableView runs the registered listeners on each received message. TableView users could utilize TableView Listeners for event-sourcing applications.
+
+### How to Register TableView Listeners
+
+Currently, `TableView#foreachAndListen` and `TableView#listen` interfaces are used to register listeners.   
+
+# Motivation
+issue : https://github.com/apache/pulsar/issues/23158
+
+As reported in the above issue, currently, the TableView listener registration methods, `foreachAndListen` and `listen`, are being called after `TableView#start()`. This timing issue can result in some messages being missed or not handled if the listeners are registered too late.
+
+Additionally, the current implementation of these listeners does not distinguish between existing messages (those present before the TableView started) and tail (future) messages (those received after the TableView started). Some business logic may require different handling for these two types of messages, but the current design does not support this differentiation.
+
+
+
+# Goals
+
+## In Scope
+
+- Support specifying TableView Listeners in TableView builder and constructor, so that TableView can register listeners as part of instance init and accordingly call them for all messages.
+- Support specifying TableView Listeners for existing and tail messages separately.
+- This design is based on Pulsar-Java-Client.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+Add followings in TableViewBuilder.java
+```java
+    /**
+     * Set the message listeners.
+     * If {@link TableViewBuilder#existingMessageListeners} are not specified, 
+     * these listeners are used for both
+     * existing and tailing(future) messages in the topic.
```

Comment:
I don't think that it's a great idea to use this for both types. If someone wants the listener on both types, why not just register it for the other type of listener separately?

Reply:
That will require setting the same listener for both, which I think is redundant.

Reply:
I dont have strong opinion here, but I thought that we want the client design for less code.

Reply:
Alternatively, we can introduce an enum TableViewMessageType{
Existing,
Tail
}

And expose this Api only 

TableViewBuilder listeners(TableViewMessageType type, BiConsumer<String, T>... listeners)